packages/langium/test/grammar/grammar-validator.test.ts

```diff
@@ -698,6 +698,233 @@ describe('Property type is not a mix of cross-ref and non-cross-ref types.', ()
+            Person: 'person' name=ID ;
+        `));
+        expect(validation.diagnostics.length).toBe(2);
+        expect(validation.diagnostics[0].message).toBe(getMessage('persons'));
+        expect(validation.diagnostics[1].message).toBe(getMessage('persons'));
+    });
+
+    test('multiple, nested optional assignments', async () => {
+        const validation = await validate(getGrammar(`
+            entry Model:
+                persons=Person (',' persons=Person (',' persons=Person )?)?;
+            Person: 'person' name=ID ;
+        `));
+        expect(validation.diagnostics.length).toBe(3);
+        expect(validation.diagnostics[0].message).toBe(getMessage('persons'));
+        expect(validation.diagnostics[1].message).toBe(getMessage('persons'));
+        expect(validation.diagnostics[2].message).toBe(getMessage('persons'));
+    });
+
+    test('multiple, nested mandatory assignments', async () => {
+        const validation = await validate(getGrammar(`
+            entry Model:
+                persons=Person (',' persons=Person (',' persons=Person ));
+            Person: 'person' name=ID ;
+        `));
+        expect(validation.diagnostics.length).toBe(3);
+        expect(validation.diagnostics[0].message).toBe(getMessage('persons'));
+        expect(validation.diagnostics[1].message).toBe(getMessage('persons'));
+        expect(validation.diagnostics[2].message).toBe(getMessage('persons'));
+    });
+
+    test('fragments: 2nd critical assignment is located in a fragment', async () => {
+        const validation = await validate(getGrammar(`
+            entry Model:
+                persons=Person ';' Assign;
+            fragment Assign:
+                ',' persons=Person;
+            Person: 'person' name=ID ;
+        `));
+        expect(validation.diagnostics.length).toBe(2);
+        expect(validation.diagnostics[0].message).toBe(getMessage('persons'));
+        expect(validation.diagnostics[1].message).toBe(getMessage('persons'));
+    });
+
+    test('fragments: all assignments are located in a fragment', async () => {
+        const validation = await validate(getGrammar(`
+            entry Model:
+                Assign ';' Assign;
+            fragment Assign:
+                ',' persons=Person;
+            Person: 'person' name=ID ;
+        `));
+        expect(validation.diagnostics.length).toBe(1);
+        expect(validation.diagnostics[0].message).toBe(getMessage('persons'));
+    });
+
+    test('fragments: assignment in looped fragment', async () => {
+        const validation = await validate(getGrammar(`
+            entry Model:
+                Assign*;
+            fragment Assign:
+                ',' persons=Person;
+            Person: 'person' name=ID ;
+        `));
+        expect(validation.diagnostics.length).toBe(1);
+        expect(validation.diagnostics[0].message).toBe(getMessage('persons'));
+    });
+
+    test('no problem: fragments in alternatives', async () => {
```

Comment:
What are about fragments X that are twice in one rule alternative and only once in another :)? Then the user is forced to use a list for the second case.

Reply:
That is an interesting case and you are completely right!

In my eyes, using a list in the 2nd case as well is better than loosing a value in the 1st case.
Therefore, I would stay with the current validation.

Nevertheless, I added this as an additional test case.

What do you think?